### DISCUSSION TITLE
bpftrace vs perf

### DISCUSSION THREAD

**Discussion by yugeshk** [bpftrace/bpftrace#2078] [Q&A]
Hi,
I am new to kernel tracing and have been comparing perf and bpftrace. I have two questions - 

1. I noticed that the list of `kprobes` that `bpftrace` and `perf` mention are different. I don't understand how this can be, since I thought that these points were put in the kernel? (see screenshots)- 

![image](https://user-images.githubusercontent.com/31157397/143570831-6d479c9e-f915-4d7b-a1bc-fa0950a8cbad.png)
![image](https://user-images.githubusercontent.com/31157397/143570881-8b64c5bb-cb92-4b15-ae75-714771d246e7.png)

2. I am interested in tracing the `bpf_map` functions for BPF programs attached to some event (e.g., `bpf_map_update_elem` which is called as part of a bpf program that I attached in the kernel for socket filtering). I'm not able to get any calls to such functions by using `bpftrace` -- any insights on how you would go about doing this would help a lot. 

Thanks!

**Comment by fbs** [comment]
your `perf` command lists tracepoints and events, your `bpftrace` one only lists kprobes. You might want to give `perf probe -l` a go

**Reply by yugeshk** [reply]
@fbs Heres the output of `perf probe -l` - it only lists one probe that I added manually
![image](https://user-images.githubusercontent.com/31157397/143579575-e769c60d-e237-424b-a86c-f3f6552fe2a5.png)

**Reply by fbs** [reply]
ah im not that familiar with perf, but it seems like you're listing different kinds of probes between the tools which explains the diff

**Reply by yugeshk** [reply]
I see. regardless of this difference, since for `bpftrace` I am able to see `kprobe:bpf_map_update_elem` as a possible probe, shouldn't this imply that if I run any `bpftrace` program and trace the event `kprobe:bpf_map_update_elem` I should be able to see details about all calls to `bpf_map_update_elem` made from `ebpf` programs attached to various parts of the kernel?

**Reply by yugeshk** [reply]
My best guess is that with `perf` I need to manually attach a `probe` before it is available for tracing while with `bpftrace` I can attach to all locations in the kernel symbol table, and `bpftrace` attaches the probe when its called.

**Reply by fbs** [reply]
Sounds like `perf probe -l` only shows the attached ones and bpftrace shows all available kprobes, even those that no program has been attached to.
